src/Analysis/Codelyzer.Analysis/CodeGraph.cs

```diff
@@ -0,0 +1,198 @@
+﻿using Codelyzer.Analysis.Model;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Codelyzer.Analysis
+{
+    public class CodeGraph
+    {
+        public List<Node> ProjectGraph { get; set; }
+        public List<Node> ClassGraph { get; set; }
+
+        public void Initialize(List<AnalyzerResult> analyzerResults)
+        {
+            ProjectGraph = CreateProjectGraph(analyzerResults);
+            ClassGraph = CreateClassesGraph(analyzerResults);
+        }
+        private List<Node> CreateProjectGraph(List<AnalyzerResult> analyzerResults)
+        {
+            var graph = new List<Node>();
+
+            // Add Nodes
+            foreach (var analyzerResult in analyzerResults)
+            {
+                var projectResult = analyzerResult.ProjectResult;
+                var projectNode = new Node() { NodeType = NodeType.Project, Name = projectResult.ProjectName, Identifier = projectResult.ProjectFilePath };
+                graph.Add(projectNode);
+            }
+
+            // Add Edges
+            foreach (var node in graph)
+            {
+                var project = analyzerResults.FirstOrDefault(a => a.ProjectResult.ProjectFilePath == node.Identifier);
+                var projectReferences = project.ProjectResult?.ExternalReferences?.ProjectReferences;
+
+                projectReferences.ForEach(projectReference => {
+                    var targetNode = graph.FirstOrDefault(n => n.Identifier == projectReference.AssemblyLocation);
+                    node.Edges.Add(new Edge() { EdgeType = EdgeType.ProjectReference, TargetNode = targetNode, SourceNode = node });
+                });
+            }
+
+            return graph;
+        }
+        private List<Node> CreateClassesGraph(List<AnalyzerResult> analyzerResults)
```

Comment:
I updated the algorithm:
- Instead of using the All extension method (e.g. AllInvocationExpressions), we now navigate the nodes only once and add to their buckets - All will navigate all the structure each time it's invoked
- Created HashSets for nodes for faster retrievals
- Filtered the child nodes using the code namespaces
Let me know your thoughts.